base/heartbeat_test.go

```diff
@@ -75,6 +76,90 @@ func TestNewCouchbaseHeartbeater(t *testing.T) {
 	assert.True(t, heartbeatStoppedHandler2.staleDetectCount >= 1 || heartbeatStoppedHandler3.staleDetectCount >= 1,
 		fmt.Sprintf("Expected stale detection counts (1) not found in either handler2 (%d) or handler3 (%d)", heartbeatStoppedHandler2.staleDetectCount, heartbeatStoppedHandler3.staleDetectCount))
 
+	// Validate current node list
+	activeNodes, err := node2.GetNodeList()
+	require.NoError(t, err, "Error getting node list")
+	assert.Equal(t, 2, len(activeNodes))
+	assert.NotContains(t, activeNodes, "node1")
+	assert.Contains(t, activeNodes, "node2")
+	assert.Contains(t, activeNodes, "node3")
+
+	// Stop heartbeaters
+	node2.Stop()
+	node3.Stop()
+
+}
+
+// TestCouchbaseHeartbeaterUsingViews tests the same functionality as TestCouchbaseHeartbeater,
+// but using ViewBackedNodeListHandlers.
+func TestCouchbaseHeartbeaterUsingViews(t *testing.T) {
```

Comment:
Could write a single test to try both implementations:

```golang
// TestCouchbaseHeartbeaters tries each handler implementation to simulate three nodes.
// The minimum time window for failed node detection is 2 seconds, based on
// Couchbase Server's minimum document expiry TTL of 1 second, so retry polling is required.
func TestCouchbaseHeartbeaters(t *testing.T) {

	if UnitTestUrlIsWalrus() {
		t.Skip("This test won't work under walrus - no expiry, required for heartbeats")
	}

	if testing.Short() {
		t.Skip("Skipping heartbeattest in short mode")
	}

	const keyprefix = "hbtest"

	heartbeaters := []struct {
		name                      string
		nodeSetHandlerConstructor func(bucket Bucket) HeartbeatNodeSetHandler
	}{
		{
			"documentBackedNodeListHandler",
			func(bucket Bucket) HeartbeatNodeSetHandler {
				handler, _ := NewDocumentBackedNodeListHandler(bucket, keyprefix)
				return handler
			},
		},
		{
			"viewBackedNodeListHandler",
			func(bucket Bucket) HeartbeatNodeSetHandler {
				handler, _ := NewViewBackedNodeListHandler(bucket, keyprefix)
				return handler
			},
		},
	}

	for _, h := range heartbeaters {
		t.Run(h.name, func(tt *testing.T) {

			testBucket := GetTestBucket(tt)
			defer testBucket.Close()

			// Create three heartbeaters (representing three nodes)
			handler1 := h.nodeSetHandlerConstructor(testBucket)
			assert.NotNil(tt, handler1)
			node1, err := NewCouchbaseHeartbeater(testBucket, keyprefix, "node1", handler1)
			assert.NoError(tt, err)
			handler2 := h.nodeSetHandlerConstructor(testBucket)
			assert.NotNil(tt, handler2)
			node2, err := NewCouchbaseHeartbeater(testBucket, keyprefix, "node2", handler2)
			assert.NoError(tt, err)
			handler3 := h.nodeSetHandlerConstructor(testBucket)
			assert.NotNil(tt, handler3)
			node3, err := NewCouchbaseHeartbeater(testBucket, keyprefix, "node3", handler3)
			assert.NoError(tt, err)

			assert.NoError(tt, node1.StartSendingHeartbeats(1))
			assert.NoError(tt, node2.StartSendingHeartbeats(1))
			assert.NoError(tt, node3.StartSendingHeartbeats(1))

			heartbeatStoppedHandler1 := &TestHeartbeatStoppedHandler{handlerID: "handler1"}
			heartbeatStoppedHandler2 := &TestHeartbeatStoppedHandler{handlerID: "handler2"}
			heartbeatStoppedHandler3 := &TestHeartbeatStoppedHandler{handlerID: "handler3"}

			assert.NoError(tt, node1.StartCheckingHeartbeats(1000, heartbeatStoppedHandler1))
			assert.NoError(tt, node2.StartCheckingHeartbeats(1000, heartbeatStoppedHandler2))
			assert.NoError(tt, node3.StartCheckingHeartbeats(1000, heartbeatStoppedHandler3))

			// Wait for node1 to start running (and persist initial heartbeat docs) before stopping
			retryUntilFunc := func() bool {
				return node1.checkCount > 0 && node1.sendCount > 0
			}
			testRetryUntilTrue(t, retryUntilFunc)
			assert.True(tt, node1.checkCount > 0)
			assert.True(tt, node1.sendCount > 0)

			// Stop node 1
			node1.Stop()

			// Wait for another node to detect node1 has stopped sending heartbeats
			retryUntilFunc = func() bool {
				return heartbeatStoppedHandler2.staleDetectCount >= 1 || heartbeatStoppedHandler3.staleDetectCount >= 1
			}
			testRetryUntilTrue(tt, retryUntilFunc)

			// Validate that at least one node detected the stopped node 1
			assert.True(tt, heartbeatStoppedHandler2.staleDetectCount >= 1 || heartbeatStoppedHandler3.staleDetectCount >= 1,
				fmt.Sprintf("Expected stale detection counts (1) not found in either handler2 (%d) or handler3 (%d)", heartbeatStoppedHandler2.staleDetectCount, heartbeatStoppedHandler3.staleDetectCount))

			// Validate current node list
			activeNodes, err := node2.GetNodeList()
			require.NoError(tt, err, "Error getting node list")
			assert.Equal(tt, 2, len(activeNodes))
			assert.NotContains(tt, activeNodes, "node1")
			assert.Contains(tt, activeNodes, "node2")
			assert.Contains(tt, activeNodes, "node3")

			// Stop heartbeaters
			node2.Stop()
			node3.Stop()
		})
	}

}
```

Reply:
Yeah, pure laziness on my part.  I went back and made this change.